include/inexor/vulkan-renderer/frame_graph.hpp

```diff
@@ -0,0 +1,419 @@
+    PhysicalGraphicsStage(const PhysicalGraphicsStage &) = delete;
+    PhysicalGraphicsStage(PhysicalGraphicsStage &&) = delete;
+    ~PhysicalGraphicsStage() override;
+
+    PhysicalGraphicsStage &operator=(const PhysicalGraphicsStage &) = delete;
+    PhysicalGraphicsStage &operator=(PhysicalGraphicsStage &&) = delete;
+};
+
+class FrameGraph {
+private:
+    VkDevice m_device;
+    VkCommandPool m_command_pool;
+    VmaAllocator m_allocator;
+    const wrapper::Swapchain &m_swapchain;
+    std::shared_ptr<spdlog::logger> m_log;
+
+    // NOTE: unique_ptr must be used as Render* is just the base class
+    std::vector<std::unique_ptr<RenderResource>> m_resources;
+    std::vector<std::unique_ptr<RenderStage>> m_stages;
+
+    // Stage execution order
+    std::vector<RenderStage *> m_stage_stack;
+
+    // Resource to physical resource map
+    std::unordered_map<const RenderResource *, std::unique_ptr<PhysicalResource>> m_resource_map;
+
+    // Stage to physical stage map
+    std::unordered_map<const RenderStage *, std::unique_ptr<PhysicalStage>> m_stage_map;
+
+    // Helper function used to create a physical resource during frame graph compilation
+    // TODO(): Use concepts when we switch to C++ 20
+    template <typename T, typename... Args, std::enable_if_t<std::is_base_of_v<PhysicalResource, T>, int> = 0>
+    T *create(const RenderResource *resource, Args &&... args) {
+        auto ptr = std::make_unique<T>(std::forward<Args>(args)...);
+        auto *ret = ptr.get();
+        m_resource_map.emplace(resource, std::move(ptr));
+        return ret;
+    }
+
+    // Helper function used to create a physical stage during frame graph compilation
+    // TODO(): Use concepts when we switch to C++ 20
+    template <typename T, typename... Args, std::enable_if_t<std::is_base_of_v<PhysicalStage, T>, int> = 0>
+    T *create(const RenderStage *stage, Args &&... args) {
+        auto ptr = std::make_unique<T>(std::forward<Args>(args)...);
+        auto *ret = ptr.get();
+        m_stage_map.emplace(stage, std::move(ptr));
+        return ret;
+    }
+
+    // Physical resources
+    void build_image(const TextureResource *, PhysicalImage *, VmaAllocationCreateInfo *);
+    void build_image_view(const TextureResource *, PhysicalImage *);
+
+    // Physical stages
+    void build_render_pass(const GraphicsStage *, PhysicalGraphicsStage *);
+    void build_graphics_pipeline(const GraphicsStage *, PhysicalGraphicsStage *);
+    void alloc_command_buffers(const RenderStage *, PhysicalStage *);
+    void record_command_buffers(const RenderStage *, PhysicalStage *);
+
+public:
+    FrameGraph(VkDevice device, VkCommandPool command_pool, VmaAllocator allocator,
+               const wrapper::Swapchain &swapchain);
+
+    /// @brief Adds either a render resource or render stage to the frame graph
+    /// @return A mutable reference to the just-added resource or stage
+    template <typename T, typename... Args>
```

Comment:
You nailed [rule P.5 of C++ core guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rp-compile-time) here: Prefer compile-time checking to run-time checking.
That's amazing use of modern C++ there!